src/main/java/edu/kit/compiler/optimizations/Util.java

```diff
@@ -236,4 +238,36 @@ public void visitBlock(Block block) {
 
     }
 
+    /**
+     * Call `func` for each non-Bad direct control flow predecessor of `block`.
+     */
+    public static void forEachPredBlock(Block block, ObjIntConsumer<Block> func) {
+        for (int i = 0; i < block.getPredCount(); ++i) {
+            var pred = block.getPred(i);
+            if (pred.getOpCode() != ir_opcode.iro_Bad) {
+                func.accept((Block) pred.getBlock(), i);
+            }
+        } 
+    }
+
+    /**
+     * Returns a map containing a map of blocks to a list of contained nodes.
+     */
+    public static Map<Block, List<Node>> getNodesPerBlock(Graph graph) {
```

Comment:
`Util.BlockNodeMapper`?

Reply:
Yes, good point, I may be able to reuse that. It is not the exact same thing, the `BlockNodeMapper` includes the `Block` itself in the `List`. I'm not sure if this is relevant for my use case.

Reply:
Does it?

Reply:
Yes, `defaultVisit` is also called for blocks.

Reply:
Funny, I never realized that :D It doesn't matter for the `BlockNodeMapper` though, since `block.getBlock()` returns `null`.